File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/DateButton.java
package org.totschnig.myexpenses.ui;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.ContextThemeWrapper;
import android.widget.CalendarView;

import org.threeten.bp.LocalDate;
import org.threeten.bp.format.DateTimeFormatter;
import org.threeten.bp.format.FormatStyle;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.util.Utils;

import java.util.Locale;

import icepick.State;

/**
 * A button that opens DateDialog, and stores the date in its state
 */
public class DateButton extends ButtonWithDialog {
  @State
  LocalDate date;
  private DateTimeFormatter dateFormat = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT);

  public Dialog onCreateDialog() {
    requireDate();
    boolean brokenSamsungDevice = isBrokenSamsungDevice();
    final Context base = getContext();
    @SuppressLint("InlinedApi")
    Context context = brokenSamsungDevice ?
        new ContextThemeWrapper(base,
            base instanceof ProtectedFragmentActivity && ((ProtectedFragmentActivity) base).getThemeType() == ProtectedFragmentActivity.ThemeType.dark ?
                android.R.style.Theme_Holo_Dialog : android.R.style.Theme_Holo_Light_Dialog) :
        base;
    int yearOld = date.getYear();
    int monthOld = date.getMonthValue() - 1;
    int dayOld = date.getDayOfMonth();
    /**
     * listens on changes in the date dialog and sets the date on the button
     */
    DatePickerDialog.OnDateSetListener mDateSetListener =
        (view, year, monthOfYear, dayOfMonth) -> {
          if (yearOld != year ||
              monthOld != monthOfYear  ||
              dayOld != dayOfMonth) {
            setDate(LocalDate.of(year, monthOfYear + 1, dayOfMonth));
            getHost().onValueSet(this);
          }
        };
    DatePickerDialog datePickerDialog = new DatePickerDialog(context, mDateSetListener,
        yearOld, monthOld, dayOld);
    if (brokenSamsungDevice) {
      datePickerDialog.setTitle("");
      datePickerDialog.updateDate(yearOld, monthOld, dayOld);
    }
    if (PrefKey.GROUP_WEEK_STARTS.isSet()) {
      int startOfWeek = Utils.getFirstDayOfWeekFromPreferenceWithFallbackToLocale(Locale.getDefault());
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
        datePickerDialog.getDatePicker().setFirstDayOfWeek(startOfWeek);
      } else {
        try {
          setFirstDayOfWeek(datePickerDialog, startOfWeek);
        } catch (UnsupportedOperationException e) {/*Nothing left to do*/}
      }
    }

    return datePickerDialog;
  }

  private void requireDate() {
    if (date == null) {
      date = LocalDate.now();
    }
  }

  private void setFirstDayOfWeek(DatePickerDialog datePickerDialog, int startOfWeek) {
    CalendarView calendarView = datePickerDialog.getDatePicker().getCalendarView();
    calendarView.setFirstDayOfWeek(startOfWeek);
  }

  private static boolean isBrokenSamsungDevice() {
    return (Build.MANUFACTURER.equalsIgnoreCase("samsung")
        && isBetweenAndroidVersions(
        Build.VERSION_CODES.LOLLIPOP,
        Build.VERSION_CODES.LOLLIPOP_MR1));
  }

  private static boolean isBetweenAndroidVersions(int min, int max) {
    return Build.VERSION.SDK_INT >= min && Build.VERSION.SDK_INT <= max;
  }

  public void setDate(LocalDate localDate) {
    date = localDate;
    update();
  }

  @NonNull
  public LocalDate getDate() {
    requireDate();
    return date;
  }

  public DateButton(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
  }

  @Override
  protected void update() {
    requireDate();
    setText(date.format(dateFormat));
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/ui/DateButton.kt
package org.totschnig.myexpenses.ui

import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.core.widget.TextViewCompat
import com.evernote.android.state.State
import com.google.android.material.datepicker.MaterialDatePicker
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.util.epochMillis2LocalDate
import org.totschnig.myexpenses.util.getDateTimeFormatter
import org.totschnig.myexpenses.util.ui.UiUtils
import org.totschnig.myexpenses.util.ui.preferredDatePickerBuilder
import org.totschnig.myexpenses.util.ui.resolveThemeColor
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle


/**
 * A button that opens DateDialog, and stores the date in its state
 */

class DateButton @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : ButtonWithDialog<MaterialDatePicker<Long>>(context, attrs, defStyleAttr) {

    private var lastTouchDownX: Float? = null
    private val marginTouchWidth = UiUtils.dp2Px(36F, resources)

    @State
    @JvmField
    var date: LocalDate = LocalDate.now()

    private val formatter: DateTimeFormatter = getDateTimeFormatter(context, shortYear = true)

    init {
        this.setCompoundDrawablesRelativeWithIntrinsicBounds(
            R.drawable.ic_chevron_start,
            0,
            R.drawable.ic_chevron_end,
            0
        )
        compoundDrawablePadding = 0
        TextViewCompat.setCompoundDrawableTintList(
            this,
            ColorStateList.valueOf(
                resolveThemeColor(
                    getContext(),
                    androidx.appcompat.R.attr.colorPrimary
                )
            )
        )
        val horizontalPadding = 0
        setPaddingRelative(horizontalPadding, paddingTop, horizontalPadding, paddingBottom)
        //noinspection ClickableViewAccessibility
        setOnTouchListener { _, motionEvent ->
            if (motionEvent.actionMasked == MotionEvent.ACTION_DOWN) {
                lastTouchDownX = motionEvent.x
            }
            false
        }
    }


    override fun onClick() {
        lastTouchDownX?.let {
            when {
                it <= marginTouchWidth -> {
                    previousDay()
                }

                it >= width - marginTouchWidth -> {
                    nextDay()
                }

                else -> {
                    super.onClick()
                }
            }
        } ?: kotlin.run {
            super.onClick()
        }
    }

    private fun previousDay() {
        setDateInternal(date.minusDays(1))
    }

    private fun nextDay() {
        setDateInternal(date.plusDays(1))
    }

    private fun setDateInternal(localDate: LocalDate) {
        setDate(localDate)
        host.onValueSet(this)
    }

    fun setDate(localDate: LocalDate) {
        date = localDate
        update()
    }

    override val fragmentTag: String
        get() = "date_button"

    override fun buildDialog() = preferredDatePickerBuilder(context)
        .setSelection(
            ZonedDateTime.of(date.atStartOfDay(), ZoneId.of("UTC")).toEpochSecond() * 1000
        )
        .build()

    override fun attachListener(dialogFragment: MaterialDatePicker<Long>) {
        dialogFragment.addOnPositiveButtonClickListener {
            setDateInternal(epochMillis2LocalDate(it, ZoneId.of("UTC")))
            context.injector.prefHandler()
                .putInt(PrefKey.DATE_PICKER_INPUT_MODE, dialogFragment.inputMode)
        }
        dialogFragment.addOnDismissListener {
            dialogShown = false
        }
    }

    override fun update() {
        text = date.format(formatter)
        contentDescription = date.format(DateTimeFormatter.ofLocalizedDate(FormatStyle.LONG))
    }

    fun overrideText(text: CharSequence) {
        this.text = text
        setCompoundDrawables(null, null, null, null)
    }
}
